src/runtime/bind_stream.rs

```diff
@@ -0,0 +1,177 @@
+mod waker;
+use waker::*;
+
+use super::{Revision, Runtime};
+use futures::{
+    channel::mpsc,
+    stream::{Stream, StreamExt},
+    task::{noop_waker, waker, LocalSpawn},
+};
+use std::{
+    pin::Pin,
+    task::{Context as FutContext, Poll},
+};
+
+/// A [`Runtime`] that is bound with a particular root function.
+///
+/// If running in a context with an async executor, can be consumed as a
+/// [`futures::Stream`] of [`crate::runtime::Revision`]s in order to provide
+/// the [`super::Runtime`] with a [`std::task::Waker`].
+pub struct BindStream<Root> {
+    inner: Runtime,
+    root: Root,
+    changes_receiver: mpsc::Receiver<()>,
```

Comment:
Could this be something like `Arc<AtomicBool>` instead? If multiple state updates occur before the stream is polled again we shouldn't run multiple times, right?

Reply:
This was my first attempt. The problem was about wakers somehow getting invalidated after storing them, before some change occurs, therefore calling noop on wake. I looked up example of `Stream` implementation that was just a wrapper of another stream/future. So i scrapped my initial attempt and just wrapped mpsc receiver as it implements `Stream` trait. And it worked! The next logical step would be to see how mpsc implemented and use that... buuuut i don't see if there can be any significant performance gain while sacrificing some maintainability and adding more code. Either way there will be some indirection because wrapping callback inside a waker, to send message through channel and give receiver control over waker or to update atomic bool then calling wrapped wake.

Reply:
> And we use bounded mpsc channel practically with a buffer for one message, ignoring buffer full error and not sending a new message (should add test for this)

Ah, I missed this in my first pass. This makes sense and is fine as long as it's documented.